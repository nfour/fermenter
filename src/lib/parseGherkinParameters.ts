import { IGherkinAstExamples, IGherkinAstStep, IGherkinParameter, IMatch } from '../types';
import { ExpressionMatcher } from './ExpressionMatcher';
import { GherkinTableReader } from './GherkinTableReader';

export function parseGherkinParameters (stepOrExamples: IGherkinAstStep | IGherkinAstExamples, match: IMatch) {
  const text = stepOrExamples.type === 'Step'
    ? stepOrExamples.text
    : stepOrExamples.name;

  const matcher = ExpressionMatcher(match);
  const matches = matcher.match(text) || [];
  const simpleParams = matches.map((arg) => arg.getValue(null));

  const paramSource = stepOrExamples.type === 'Step'
    ? stepOrExamples.argument
    : stepOrExamples;

  const argumentParam = parseGherkinForParameter(paramSource);

  const params = argumentParam
    ? [...simpleParams, argumentParam]
    : simpleParams;

  return params;
}

function parseGherkinForParameter (arg: IGherkinAstStep['argument'] | IGherkinAstExamples): undefined | IGherkinParameter {
  if (!arg) { return; }

  if (arg.type === 'DataTable') {
    return GherkinTableReader({ rows: arg.rows });
  }

  if (arg.type === 'Examples') {
    const rows = [
      arg.tableHeader,
      ...arg.tableBody || [],
    ];

    return GherkinTableReader({ rows });
  }

  if (arg.type === 'DocString') {
    return arg.content;
  }
}
