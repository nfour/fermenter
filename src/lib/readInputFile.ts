declare const __non_webpack_require__: any;
declare const __webpack_require__: any;

export function readInputFile ({ filePath, testFilePath = '' }: { filePath: string, testFilePath?: string }) {
  const dRequire: typeof require = typeof __webpack_require__ === 'function' ? __non_webpack_require__ : require;

  const { dirname, isAbsolute, resolve }: typeof import ('upath') = dRequire('upath');
  const { readFileSync }: typeof import ('fs') = dRequire('fs');

  if (isAbsolute(filePath)) {
    return readFileSync(filePath, 'utf8');
  }

  const testDirectory = testFilePath && dirname(testFilePath);
  const finalPath = resolve(testDirectory, filePath);

  return readFileSync(finalPath, 'utf8');
}
