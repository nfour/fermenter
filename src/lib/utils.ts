export type IDeferredPromise<T = any> = Promise<T> & { resolve: () => any };

export function deferredPromise<T> (): IDeferredPromise<T> {
  let resolver: undefined | (() => any);

  const promise = <IDeferredPromise<T>> new Promise<T>((resolve) => resolver = resolve);
  promise.resolve = resolver!;

  return promise;
}
