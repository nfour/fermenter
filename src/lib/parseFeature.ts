import * as errorParser from 'error-stack-parser';
import * as Gherkin from 'gherkin';
import * as isValidPath from 'is-valid-path';
import { isAbsolute, toUnix } from 'upath';

import { FeatureBuilder } from '../FeatureBuilder';
import { IGherkinAst } from '../types';
import { readInputFile } from './readInputFile';

export interface IGherkinParserOutput {
  ast: IGherkinAst;
  text: string;
  featureBuilder: FeatureBuilder;
}

export interface IGherkinParserConfig {
  feature: string;
  stackIndex?: number;
}

/** Parses a feature file from a relative filePath */
export function parseFeature ({ feature, stackIndex = 1 }: IGherkinParserConfig): IGherkinParserOutput {
  const text = (() => {
    // Is a feature file body
    if (!isValidPath(feature)) { return feature; }

    const filePath = toUnix(feature);

    const testFilePath = (() => {
      if (isAbsolute(feature)) { return; }

      const error = new Error();
      /** Increment stackIndex due to the nested IIFE's it's within */
      return errorParser.parse(error)[stackIndex + 2].fileName;
    })();

    return readInputFile({ filePath, testFilePath });
  })();

  const parser = new Gherkin.Parser();
  const ast: IGherkinAst = parser.parse(text);

  const featureBuilder = new FeatureBuilder(ast);

  return { ast, text, featureBuilder };
}
