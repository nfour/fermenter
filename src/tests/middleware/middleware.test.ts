import { resolve } from 'path';

import { executeTest } from '../lib';

jest.setTimeout(10000);

describe('Middleware hook tests', () => {
  it('middlewareHookTest', async () => {
    const output = await executeTest(resolve(__dirname, './middlewareHookTest.ts'));

    expect(output).toMatchSnapshot();
  });
});
