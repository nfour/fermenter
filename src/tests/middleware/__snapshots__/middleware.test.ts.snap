// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Middleware hook tests middlewareHookTest 1`] = `
" PASS  src/tests/middleware/middlewareHookTest.ts
  Feature: Middleware COR for steps, globally
    Scenario: I have a scenario
      ✓ Given: Something is set by a middleware
      ✓ Then: Check that Something is true

Test Suites: 1 passed, 1 total
Tests:       2 passed, 2 total"
`;
